wsl.go

```diff
@@ -370,9 +399,13 @@ func (p *Processor) parseBlockStatements(statements []ast.Stmt) {
 		case *ast.ExprStmt:
 			switch previousStatement.(type) {
 			case *ast.DeclStmt, *ast.ReturnStmt:
-				p.addError(t.Pos(), "expressions should not be cuddled with declarations or returns")
+				if cuddledWithLastStmt {
```

Comment:
Hmm, sorry, I realise when I see this that my feedback regarding the unncssesary type switch for all statements might've been bad. Sadly it's not sufficient to make these the only `if cuddledWithLastStmt`.

Since none of the existing tests covers when enforcing is true, the tests didn't catch this. This is the result of running `wsl` on itself with enforcing true:

```text
cmd/wsl/main.go:43: if statements should only be cuddled with assignments
cmd/wsl/main.go:48: assignments should only be cuddled with other assignments
cmd/wsl/main.go:59: branch statements should not be cuddled if block has more than two lines
cmd/wsl/main.go:53: only one cuddle assignment allowed before range statement
cmd/wsl/main.go:68: ranges should only be cuddled with assignments used in the iteration
cmd/wsl/main.go:84: assignments should only be cuddled with other assignments
cmd/wsl/main.go:87: only one cuddle assignment allowed before range statement
cmd/wsl/main.go:95: only one cuddle assignment allowed before range statement
cmd/wsl/main.go:91: if statements should only be cuddled with assignments
cmd/wsl/main.go:100: if statements should only be cuddled with assignments used in the if statement itself
cmd/wsl/main.go:119: return statements should not be cuddled if block has more than two lines
examples/analyze/main.go:25: return statements should not be cuddled if block has more than two lines
wsl.go:160: only cuddled expressions if assigning variable or using from line above
wsl.go:163: return statements should not be cuddled if block has more than two lines
wsl.go:178: return statements should not be cuddled if block has more than two lines
wsl.go:171: if statements that check an error must be cuddled with the statement that assigned the error
wsl.go:181: assignments should only be cuddled with other assignments
wsl.go:184: only one cuddle assignment allowed before range statement
wsl.go:228: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:237: only one cuddle assignment allowed before if statement
wsl.go:244: declarations should never be cuddled
wsl.go:248: declarations should never be cuddled
wsl.go:251: declarations should never be cuddled
wsl.go:262: if statements should only be cuddled with assignments
wsl.go:270: declarations should never be cuddled
wsl.go:276: declarations should never be cuddled
wsl.go:285: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:291: if statements should only be cuddled with assignments
wsl.go:302: return statements should not be cuddled if block has more than two lines
wsl.go:295: assignments should only be cuddled with other assignments
wsl.go:320: return statements should not be cuddled if block has more than two lines
wsl.go:331: branch statements should not be cuddled if block has more than two lines
wsl.go:334: only one cuddle assignment allowed before if statement
wsl.go:339: if statements should only be cuddled with assignments
wsl.go:344: if statements should only be cuddled with assignments
wsl.go:348: if statements should only be cuddled with assignments
wsl.go:375: branch statements should not be cuddled if block has more than two lines
wsl.go:378: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:388: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:417: if statements should only be cuddled with assignments
wsl.go:425: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:434: if statements should only be cuddled with assignments
wsl.go:464: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:473: assignments should only be cuddled with other assignments
wsl.go:478: if statements should only be cuddled with assignments
wsl.go:488: if statements should only be cuddled with assignments
wsl.go:497: if statements should only be cuddled with assignments
wsl.go:507: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:513: if statements should only be cuddled with assignments
wsl.go:523: if statements should only be cuddled with assignments
wsl.go:538: if statements should only be cuddled with assignments
wsl.go:323: type switch statements should only be cuddled with variables switched
wsl.go:570: declarations should never be cuddled
wsl.go:572: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:604: only cuddled expressions if assigning variable or using from line above
wsl.go:576: type switch statements should only be cuddled with variables switched
wsl.go:613: return statements should not be cuddled if block has more than two lines
wsl.go:619: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:672: return statements should not be cuddled if block has more than two lines
wsl.go:678: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:754: return statements should not be cuddled if block has more than two lines
wsl.go:760: type switch statements should only be cuddled with variables switched
wsl.go:781: return statements should not be cuddled if block has more than two lines
wsl.go:800: return statements should not be cuddled if block has more than two lines
wsl.go:811: return statements should not be cuddled if block has more than two lines
wsl.go:817: only one cuddle assignment allowed before range statement
wsl.go:823: ranges should only be cuddled with assignments used in the iteration
wsl.go:829: return statements should not be cuddled if block has more than two lines
wsl.go:867: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:871: assignments should only be cuddled with other assignments
wsl.go:875: only one cuddle assignment allowed before if statement
wsl.go:879: declarations should never be cuddled
wsl.go:897: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:902: ranges should only be cuddled with assignments used in the iteration
wsl.go:908: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:938: return statements should not be cuddled if block has more than two lines
wsl.go:916: if statements should only be cuddled with assignments used in the if statement itself
wsl.go:945: type switch statements should only be cuddled with variables switched
wsl.go:955: assignments should only be cuddled with other assignments
wsl.go:957: declarations should never be cuddled
wsl.go:988: assignments should only be cuddled with other assignments
wsl.go:971: ranges should only be cuddled with assignments used in the iteration
wsl.go:993: assignments should only be cuddled with other assignments
```

Reply:
Oooh yeah, I had that defaulted to On for most of my original testing, that must be the difference. No problem, I'll revert that. Thanks for the catch.

Reply:
This seems to fix the issue:

```diff
diff --git a/wsl.go b/wsl.go
index ecc1577..0bc82c4 100644
--- a/wsl.go
+++ b/wsl.go
@@ -320,12 +320,20 @@ func (p *Processor) parseBlockStatements(statements []ast.Stmt) {
 			return false
 		}
 
+		if _, ok := stmt.(*ast.IfStmt); !ok {
+			if !cuddledWithLastStmt {
+				continue
+			}
+		}
+
 		switch t := stmt.(type) {
 		case *ast.IfStmt:
-			checkingErr := atLeastOneInListsMatch(rightAndLeftHandSide, p.config.ErrorVariableNames)
-			if !cuddledWithLastStmt && checkingErr {
-				if atLeastOneInListsMatch(assignedOnLineAbove, p.config.ErrorVariableNames) {
-					p.addError(t.Pos(), "if statements that check an error must be cuddled with the statement that assigned the error")
+			if !cuddledWithLastStmt {
+				checkingErr := atLeastOneInListsMatch(rightAndLeftHandSide, p.config.ErrorVariableNames)
+				if checkingErr {
+					if atLeastOneInListsMatch(assignedOnLineAbove, p.config.ErrorVariableNames) {
+						p.addError(t.Pos(), "if statements that check an error must be cuddled with the statement that assigned the error")
+					}
 				}
 
 				continue
```

However, the error left is quite interesting:

```text
wsl.go:171: if statements that check an error must be cuddled with the statement that assigned the error
```

Which is for this line:

```go
file, err := parser.ParseFile(fileSet, filename, data, parser.ParseComments)

// If the file is not parsable let's add a syntax error and move on.
if err != nil {
    p.result = append(p.result, Result{
        FileName:   filename,
        LineNumber: 0,
        Reason:     fmt.Sprintf("invalid syntax, file cannot be linted (%s)", err.Error()),
    })

    return
}
```

I guess it's just a case where one would avoid this kind of behaviour :)

Reply:
@JDiPierro Don't rever, I think it looks good! See my proposal above which resolves the issue! :)

Reply:
Although this makes me think that at least one non-negative test should be added. That is, something that has enforced set to true but doesn't violate the case. Maybe a few regular if-statements and a return statement. You can add them all in the same tests as long as they don't report any false positives :)

Reply:
All set, I re-added the check that lets us bail early if we're not cuddled and not checking an error.

```
$ wsl -enforce-err-cuddling wsl.go
wsl.go:171: if statements that check an error must be cuddled with the statement that assigned the error
```

Reply:
Oops, just saw your comments above, I'll use that instead.

Reply:
All set! Added a test case that failed without the patch, applied your patch and it passed :)

That failure above seems like a real failure with `enforce-err-cuddling` enabled. I'd correct it to:

```go
// If the file is not parsable let's add a syntax error and move on.
file, err := parser.ParseFile(fileSet, filename, data, parser.ParseComments)
if err != nil {
    p.result = append(p.result, Result{
        FileName:   filename,
        LineNumber: 0,
        Reason:     fmt.Sprintf("invalid syntax, file cannot be linted (%s)", err.Error()),
    })

    return
}
```

But since the feature's off by default I didn't see a need to correct it :)

Reply:
Yeah, I agree! If one would use the new flag this should be reported so that's a great example!